### ISSUE TITLE
Raspberry 2 port

### ISSUE THREAD

**Issue by skmp** [reicast/reicast-emulator#601]
So, i'm waiting on my RaPi2, and others are already interested on this. 

Let the fun begin.

**Comment by kingosticks** [comment]
If it's any help/interest, my makefile is at https://github.com/kingosticks/reicast-emulator/blob/raspberrypi/shell/rpi/Makefile
It's probably a bit over the top but it did get something. Looking forward to seeing the progress you guys make

**Comment by skmp** [comment]
Can you try building https://github.com/reicast/reicast-emulator/tree/skmp/arm-hardfp ? It worked for me -- http://blog.nillware.com/post/92267281321/tegra-k1-fun-times

**Comment by skmp** [comment]
http://files.korny.cc/reicast/reicast-f85174c3e1b-armhf.7z - my build from tegra devboard

**Comment by kingosticks** [comment]
It built fine with the arm-hardfp branch (wasted my time a bit there rolling my own) but i got the same fatal error as I got with the build from my fork: http://dpaste.com/11GQ3VT

When I tried your tegra build I got:
>pi@raspberrypi ~ $ ./reicast.elf
./reicast.elf: /lib/arm-linux-gnueabihf/libm.so.6: version `GLIBC_2.15' not found (required by ./reicast.elf)

**Comment by skmp** [comment]
Your cross compiler seems bust, it uses newer libs that your target system. Can you test a simple hello word program?

**Comment by kingosticks** [comment]
I'm doing all my compilation on the raspberry pi. Sorry if I'm being dense
but does that explain mmap memory allocation failures?

**Comment by skmp** [comment]
*My* compiler is newer than your run-time, that's why it doesn't work. (Sorry missread that)

Do you have anything weird like SE linux or enabled? Is reicast run from a writable directory? Do you have /dev/shm available?

An strace would also be useful.

**Comment by skmp** [comment]
try running w/ root too

**Comment by kingosticks** [comment]
No SE linux or anything like that, this is stock raspbian. It is run from a writeable directory. 
>pi@raspberrypi ~ $ ls -l /dev/shm
lrwxrwxrwx 1 root root 8 Feb 10 00:21 /dev/shm -> /run/shm

And there is no difference when running using sudo. 

http://dpaste.com/025MVM8 is the strace up until the failure, then there's a segmentation fault and everything goes a bit mental. I guess this bit is the mmap failure I pasted earlier:
> write(1, "MAP 00800000 w/ 25165824\n", 25) = 25
3722  mmap2(0x56e40000, 2097152, PROT_READ, MAP_SHARED|MAP_FIXED, 3, 0x1800) = 0x56e40000
3722  mprotect(0x56e40000, 2097152, PROT_READ) = -1 EACCES (Permission denied)
3722  write(1, "MAP1 failed 13\n", 15)  = 15

**Comment by kingosticks** [comment]
Would insufficient ram cause this error? How much is required? I think I've currently got a 256/768 GPU/SYS split.

**Comment by skmp** [comment]
reicast needs around 80 megs of ram + opengl overhead, I don't think that's the issue.

I don't see a reason why mprotect fails there. the region is already mapped as PROT_READ. ?!?

**Comment by skmp** [comment]
EACCES
The prot argument specifies a protection that violates the access permission the process has to the underlying memory object.

Hmm, perhaps it doesn't like the unliked shm ?

**Comment by kingosticks** [comment]
Perhaps but why would raspbian have a problem and none of the other
systems? I've no experience with that idea, is it common?

**Comment by skmp** [comment]
Nope, but linux likes to fuck up these things on every 3rd kernel release (I already have a few workarounds for linaro and other arm boards messy mess).

on vmem.cpp try removing/commenting the "shm_unlink("/dcnzorz_mem");" line, build and run and tell me what happens :)

**Comment by kingosticks** [comment]
Commenting out that line fails the same way the first time I ran reicast. 
But on subsequent runs, now that there's already a file at /dev/shm/dcnzorz_mem causing it to create and use a dcnzorz_mem file in the working directory, things get much further (they might even work if I had started x)!

http://dpaste.com/2JEBAQ7

So rather than being the unlink, it just looks like a problem when trying to mprotect an area of /dev/shm shared memory from READ+WRITE to READ only. Sounds like a reasonable thing to do. Perhaps it's just a permission problem on my raspbian system (the permissions are viewable in the above pasted log)?

**Comment by skmp** [comment]
Yay my rapi2 is here! :3

I couldn't make sense of the shm stuff -- the exact code works fine on many other linuxes. Oh well. 

Moving on from there, I get a crash from the 3D drivers under X11. Craxy taxi boots and runs when 3d is disabled, I'm now updating the drivers...

**Comment by skmp** [comment]
New drivers make no difference. Hrfm.

**Comment by skmp** [comment]
Okay, it works now :D

**Comment by skmp** [comment]
Crazy taxi ingame ~ 75% emulation speed, just 4 fps rendering. The gpu is .. WEAK. Or software renderered.

**Comment by skmp** [comment]
I'll make a video and up sauce tomorrows, gotta *REALLY* get back to work now

**Comment by kingosticks** [comment]
Fantastic (that it's working)! Love to know what you did to make it work.

**Comment by kingosticks** [comment]
Keeping in mind that everything I know about EGL, GLES and graphics was
what I read last night, don't you have to link against /opt/vc/lib to make
use of the Rpi's videocore graphics hardware? How else could it know how to
use it? And what does update your drivers mean? With rpi-update? I think my
issue might stem from using untested dc bios files.
But Im enjoying messing around with this project even without playing
anything - so thanks!

**Comment by skmp** [comment]
Had some free time today - check the skmp/rapi2 branch. Next up: videos about it :p

**Comment by skmp** [comment]
I also hardcoded joystick/keyboard support. You can select the file via -config config:image=path/to/file

**Comment by kingosticks** [comment]
Does keyboard support work? After changing the path to `/dev/input/event0` (I made it configurable, is this something worth adding?) I see it prints the keypress events to the console but I can't actually control anything (i.e set the time in the bios). I noticed that only `TARGET_GCW0` and `TARGET_PANDORA` do anything much in `HandleKb`, is there something I'm missing?

**Comment by kingosticks** [comment]
I uncommented the pandora ifdef and used that code. Did the trick and I can hurtle around Crazy Taxi! Awesome!

With the following output, is '5.98' the frames per second and is '83.98' the emulation speed percentage? 
> newdc rel0/n - 11.91 (0.00) - 83.98 - V: 50.34 (1.00, NTSC480i59.95) R: 5.98+44.36 VTX: 0.00

**Comment by skmp** [comment]
Yeah, 83.98% is the speed, 5.98 is the actual rendered frames, 44.36 is the skipped ones because the gpu was too slow

**Comment by kingosticks** [comment]
When just sitting at the Crazy Taxi menu screen I get 19 fps and 135 sfps (!). Running htop shows 3 threads, one of which ramps up to and remains at 100% cpu utilisation. Is that what you'd expect? Is it just spinning around getting the user input?

**Comment by kingosticks** [comment]
Also, in other news, I sort of got the bottom of that `mprotext` failure. It seems that mounting /dev/shm with the `noexec` option is part of some new-fangled security improvements. If you remount it without that option then the normal `_nvmem_alloc_mem` code works just fine.  

Related bug in Chrome OS: http://code.google.com/p/nativeclient/issues/detail?id=1883

You might reasonably argue that `mprotect` is only supposed to fail if trying to set `PROT_EXEC` on a `noexec` file (which we obviously aren't). BUT we are setting `PROT_READ` and there's a very suspicious `READ_IMPLIES_EXEC` [flag](https://github.com/raspberrypi/tools/blob/master/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian/arm-linux-gnueabihf/libc/usr/include/linux/personality.h#L18), apparently for security compatibility. Strangely, `mmap` [does](https://github.com/raspberrypi/linux/blob/f105f830818642ac8bd15172bba17052c727d73a/mm/mmap.c#L1271-1278) take `noexec` into account when looking if read implies exec, and hence doesn't fail, but `mprotect` [doesn't](https://github.com/raspberrypi/linux/blob/rpi-3.18.y/mm/mprotect.c#L358-361)! That was fun.

**Comment by skmp** [comment]
Looks like I can modify the personality w/ #include <sys/personality.h> .. I'll check if that fixes the issue

**Comment by skmp** [comment]
I can verify that modifing personality does fix the issue

**Comment by skmp** [comment]
The real question is why the personality bits get set on the elf (problem w/ upstream GCC configuration?). I guess we can leave that code as extra protection :+1:
